### PR TITLE
stdlib: fix performance regression for long string appends.

### PR DESCRIPTION
**Pull request by eeckstein**
Re-wrote the inner memcpy loops so that they can be vectorized.
Also added a few inline(__always).

Since we removed some @inlineable attributes this string-append code is not code generated in the client anymore.
The code generation in the stdlib binary is different because all the precondition checks are not folded away.
Using explicit loop control statements instead of for-in-range removes the precondition-overhead for those time critical memcpy loops.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke benchmark

**Review by airspeedswift — APPROVED** [review]

File: stdlib/public/core/StringStorage.swift [hunk 1]
```diff
@@ -182,7 +182,10 @@ extension _SwiftStringStorage {
   @inlinable
   @nonobjc
   var unusedBuffer: UnsafeMutableBufferPointer<CodeUnit> {
-    return UnsafeMutableBufferPointer(start: end, count: capacity - count)
+    @inline(__always)
```
Comment (milseman) [inline stdlib/public/core/StringStorage.swift]: Why is this necessary? What model can we use to determine when we should flag something as `@inline(__always)`?

File: stdlib/public/core/StringStorage.swift [hunk 2]
```diff
@@ -182,7 +182,10 @@ extension _SwiftStringStorage {
...
+    get {
+      return UnsafeMutableBufferPointer(start: end, count: capacity - count)
```
Comment (milseman) [inline stdlib/public/core/StringStorage.swift]: `capacity &- count`?

File: stdlib/public/core/StringStorage.swift (hunk 1 above)
Comment (eeckstein) [reply stdlib/public/core/StringStorage.swift]: In this case I just wanted to ensure that this is always inlined - also with future compiler versions.
There is no general model.

File: stdlib/public/core/StringStorage.swift (hunk 2 above)
Comment (eeckstein) [reply stdlib/public/core/StringStorage.swift]: yes, makes sense

File: stdlib/public/core/StringStorage.swift (hunk 1 above)
Comment (milseman) [reply stdlib/public/core/StringStorage.swift]: https://bugs.swift.org/browse/SR-7635 😀
